### PR TITLE
nixos/modules: fix systemd start rate-limits

### PR DESCRIPTION
**Pull request by lf-**
###### Motivation for this change
These have been broken since 2016:
https://github.com/systemd/systemd/commit/f0367da7d1a61ad698a55d17b5c28ddce0dc265a
since StartLimitIntervalSec got moved into [Unit] from [Service].
StartLimitBurst has also been moved accordingly, so let's fix that one
too by adding an option ourselves.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) -- **unfortunately my computer can't run kvm so I was unable to run tests**
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lf-** [comment]
oopsie, fixed an ofborg eval error. might work now

**Comment by nh2** [comment]
This contributes to #45785, right?

Independently, could you confirm whether

> These have been broken since 2016

is really true? In https://github.com/NixOS/nixpkgs/issues/45785#issuecomment-612268015 I found that systemd did not remove `serviceConfig.StartLimitIntervalSec` -- its legacy usage is still supported, but undocumented. But I don't know if that's really the case for the burst option. As the other one is not documented, you may not get the answer from the documentation, but only from the code (or trying it).

**Comment by lf-** [comment]
I believe that's incorrect: StartLimitInterval is likely still accepted in Service (not 100% sure) but StartLimitIntervalSec is definitely not, at least according to my log lines warning on it (see the commit message and the caddy PR). StartLimitBurst I think *is* accepted there but I was touching the files anyway and it was also moved by upstream, so I fixed it as well. 

Regardless, we were relying on undocumented deprecated behaviour and I wanted to fix all of it now that I noticed it.

**Comment by lf-** [comment]
> This contributes to #45785, right?

I guess so! This fixes #45785, arguably.

**Comment by nh2** [comment]
> but StartLimitIntervalSec is definitely not

@lf- You're right, I had typod it in the linked comment and written `StartLimitIntervalSec` twice (thus making it inconsistent with the nixpkgs code I quoted there). I've fixed it now.

**Review by nh2 — APPROVED** [review]
> I guess so! This fixes #45785, arguably.

Great! +1 for me on the change (but I cannot test the change to each of the services, so just +1'ing the approach).

Optionally: Would you also be up for implementing this note from #45785 that makes any further incorrect use of the deprecated field an evaluation warning?

> * [ ] We should probably also try and make usage of `StartLimitInterval` in the old section emit a warning when it's used, as having it in the new section has [benefits](https://github.com/NixOS/nixpkgs/issues/45786#issuecomment-440091879)

**Comment by lf-** [comment]
@nh2 unfortunately I'm quite busy this next couple of weeks and probably don't have time to figure out how to do it. I doubt it's complex but I'd have to do a bit of research.

**Comment by lf-** [comment]
@nh2 I've implemented the deprecation warning as requested in 039200b.

File: nixos/modules/system/boot/systemd.nix [hunk 1]
```diff
@@ -888,9 +890,13 @@ in
...
         type = service.serviceConfig.Type or "";
         restart = service.serviceConfig.Restart or "no";
-      in optional
+        hasDeprecated = builtins.hasAttr "StartLimitInterval" service.serviceConfig;
+      in (optional
       (type == "oneshot" && (restart == "always" || restart == "on-success"))
       "Service '${name}.service' with 'Type=oneshot' cannot have 'Restart=always' or 'Restart=on-success'")
+      ++ (optional hasDeprecated ''Service '${name}.service' uses the attribute
+      StartLimitInterval in the Service section, which is deprecated. See
+      https://github.com/NixOS/nixpkgs/issues/45786.''))
```
Comment (nh2) [inline nixos/modules/system/boot/systemd.nix:899]: I don't know how warnings are rendered, but doesn't using a `''` as done here insert spaces `      ` in the middle of the message ([logic from the manual](https://nixos.org/manual/nix/stable/#idm140737322059824))?

E.g. should this be:

```
... hasDeprecated ''
  Service '${name}.service' uses the attribute
  StartLimitInterval in the Service section, which is deprecated. See
  https://github.com/NixOS/nixpkgs/issues/45786.
''
```

Comment (nh2) [reply nixos/modules/system/boot/systemd.nix:899]: I tried it, indeed the formatting is odd this way:

```
trace: warning: Service 'display-manager.service' uses the attribute
      StartLimitInterval in the Service section, which is deprecated. See
      https://github.com/NixOS/nixpkgs/issues/45786.
```

Comment (nh2) [reply nixos/modules/system/boot/systemd.nix:899]: Suggestion for the whole `warnings` block:

```nix
    warnings = concatLists (
      mapAttrsToList
        (name: service:
          let
            type = service.serviceConfig.Type or "";
            restart = service.serviceConfig.Restart or "no";
            hasDeprecated = builtins.hasAttr "StartLimitInterval" service.serviceConfig;
          in
            concatLists [
              (optional (type == "oneshot" && (restart == "always" || restart == "on-success"))
                "Service '${name}.service' with 'Type=oneshot' cannot have 'Restart=always' or 'Restart=on-success'"
              )
              (optional hasDeprecated
                "Service '${name}.service' uses the attribute 'StartLimitInterval' in the Service section, which is deprecated. See https://github.com/NixOS/nixpkgs/issues/45786."
              )
            ]
        )
        cfg.services
    );
```

Comment (nh2) [reply nixos/modules/system/boot/systemd.nix:899]: I'll force push a refactor that does this in a moment.

Comment (nh2) [reply nixos/modules/system/boot/systemd.nix]: OK done. Now it renders correctly and the diff looks much nicer.

**Review by nh2 — APPROVED** [review]
I've tested your new deprecation warnings on my system, works well. Merging.

**Comment by nh2** [comment]
I found some challenges documenting the changes in the release notes, written up in #102246.

**Comment by roberth** [comment]
Minimal 20.09 backport for compatibility: https://github.com/NixOS/nixpkgs/pull/112385

**Pull request merged**